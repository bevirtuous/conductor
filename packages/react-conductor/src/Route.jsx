import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Transition from 'react-transition-group/Transition';
import uuid from 'uuid/v4';
import Conductor from '@virtuous/conductor';

/**
 * The Route component.
 */
class Route extends Component {
  static propTypes = {
    content: PropTypes.func.isRequired,
    pattern: PropTypes.string.isRequired,
    component: PropTypes.string,
    index: PropTypes.number,
    isOpen: PropTypes.bool,
    isVisible: PropTypes.bool,
    path: PropTypes.string,
    transition: PropTypes.shape(),
  };

  static defaultProps = {
    component: 'div',
    index: null,
    isOpen: false,
    isVisible: false,
    path: null,
    transition: null,
  };

  static childContextTypes = {
    route: PropTypes.shape(),
  };

  /**
   * The constructor.
   * @param {Object} props The component props.
   */
  constructor(props) {
    super(props);

    // Generate a unique ID for this route.
    this.id = uuid();

    // Register this route with the router.
    Conductor.register(this.id, this.props.pattern);
  }

  /**
   * Creates a context about this route.
   * @returns {Object}
   */
  getChildContext() {
    return {
      route: {
        id: this.id,
      },
    };
  }

  /**
   * @param {Object} nextProps The next set of props.
   * @returns {boolean}
   */
  shouldComponentUpdate(nextProps) {
    return (
      this.props.isOpen !== nextProps.isOpen ||
      this.props.isVisible !== nextProps.isVisible
    );
  }

  /**
   * Fire the enter callback when this is the new active route.
   */
  componentDidUpdate() {
    const currentRoute = Conductor.getCurrentRoute();
    if (!this.props.transition && this.props.path === currentRoute.pathname) {
      this.routeDidEnter();
    }
  }

  /**
   * Returns a transition object based on the state of the route.
   * @returns {Object}
   */
  get transitionType() {
    const currentAction = Conductor.getCurrentAction();
    
    if (this.props.isOpen) {
      return this.props.transition.backward;
    } else if (currentAction === 'REPLACE') {
      return this.props.transition.replace;
    }

    return this.props.transition.forward;
  }

  /**
   * Inform the Conductor that the route has entered.
   */
  routeDidEnter = () => {
    const currentAction = Conductor.getCurrentAction();

    switch (currentAction) {
      case 'POP':
        Conductor.popped(this.id, this.props.path);
        break;
      case 'REPLACE':
        Conductor.replaced(this.id, this.props.path);
        break;
      default:
        Conductor.pushed(this.id, this.props.path);
        break;
    }
  }

  /**
   * Renders the component.
   * @returns {JSX}
   */
  render() {
    const {
      component: Wrapper,
      content: Content,
      index,
      isVisible,
      transition,
    } = this.props;

    if (!transition) {
      return (
        <Wrapper
          style={{
            ...(index !== null) && { zIndex: index },
            ...(!isVisible) && { display: 'none' },
          }}
        >
          <Content />
        </Wrapper>
      );
    }

    /**
     * Switch between the transitions based on whether or not
     * this route is open or if the route was replaced.
     */
    const { transitionType } = this;

    return (
      <Transition
        in={isVisible}
        timeout={transitionType.duration}
        onEntered={this.routeDidEnter}
      >
        {state => (
          <Wrapper
            style={{
              ...(index !== null) && { zIndex: index },
              ...transitionType.default,
              ...transitionType[state],
            }}
          >
            <Content />
          </Wrapper>
        )}
      </Transition>
    );
  }
}

export default Route;
