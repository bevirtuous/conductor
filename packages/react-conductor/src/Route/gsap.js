const forward = {
  duration: 375,
  default: {
    height: '100vh',
    left: '0',
    position: 'absolute',
    top: '0',
    transform: 'translate3d(100%, 0, 0)',
    width: '100%',
    willChange: 'transform',
  },
  entering: {
    pointerEvents: 'none',
    transform: 'none',
  },
  entered: {
    transform: 'none',
  },
  exiting: {
    pointerEvents: 'none',
    transform: 'translate3d(100%, 0, 0)',
  },
  exited: {
    pointerEvents: 'none',
    transform: 'translate3d(100%, 0, 0)',
  },
};

const backward = {
  duration: 375,
  default: {
    height: '100vh',
    left: '0',
    position: 'absolute',
    top: '0',
    transform: 'translate3d(-50%, 0, 0)',
    width: '100%',
    willChange: 'transform',
  },
  entering: {
    pointerEvents: 'none',
    transform: 'none',
  },
  entered: {
    transform: 'none',
  },
  exiting: {
    pointerEvents: 'none',
    transform: 'translate3d(-50%, 0, 0)',
  },
  exited: {
    pointerEvents: 'none',
    transform: 'translate3d(-50%, 0, 0)',
  },
};

const replace = {
  duration: 375,
  default: {
    height: '100vh',
    left: '0',
    position: 'absolute',
    top: '0',
    transform: 'none',
    transition: 'opacity 375ms cubic-bezier(0.25, 0.1, 0.25, 1), transform 375ms cubic-bezier(0.25, 0.1, 0.25, 1)',
    width: '100%',
    willChange: 'transform',
  },
  entering: {
    pointerEvents: 'none',
    transform: 'none',
  },
  entered: {
    transform: 'none',
  },
  exiting: {
    opacity: '0.2',
    pointerEvents: 'none',
    transform: 'translate3d(0, -10%, 0)',
  },
  exited: {
    opacity: 0,
    pointerEvents: 'none',
    transform: 'translate3d(0, -10%, 0)',
  },
};

export default {
  forward,
  backward,
  replace,
};
