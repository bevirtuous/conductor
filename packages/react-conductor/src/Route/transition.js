const transition = {
  default: {
    height: '100vh',
    left: '0',
    position: 'absolute',
    top: '0',
    width: '100%',
  },
  exiting: {
    display: 'none',
  },
  exited: {
    display: 'none',
  },
};

export default {
  forward: transition,
  backward: transition,
  replace: transition,
};
