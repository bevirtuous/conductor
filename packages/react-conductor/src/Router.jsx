import React, { Component } from 'react';
import PropTypes from 'prop-types';
import conductor from '@virtuous/conductor';
import * as events from '@virtuous/conductor-events';
import getRouteStack from '@virtuous/conductor-helpers/src/getRouteStack';

/**
 * The Router component.
 */
class Router extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    initialPath: PropTypes.string.isRequired,
    initialPattern: PropTypes.string.isRequired,
  };

  /**
   * The constructor.
   * @param {Object} props The component props.
   */
  constructor(props) {
    super(props);

    this.state = {
      routeStack: [{
        path: props.initialPath,
        pattern: props.initialPattern,
      }],
    };

    // Setup some listeners for router events.
    events.onPush(this.handleRouteChange);
    events.onPop(this.handleRouteChange);
    events.onReplace(this.handleRouteChange);

    // Push the first route.
    setTimeout(() => {
      conductor.push(props.initialPath);
    }, 0);
  }

  /**
   * Limit re-rendering unless the state actually changes.
   *
   * @param {Object} nextProps The next props.
   * @param {Object} nextState The next state.
   * @returns {boolean}
   */
  shouldComponentUpdate() {
    // TODO: Add some sort of check.
    return true;
  }

  /**
   * Find the index of the given route pattern against the currently open routes.
   * @param {string} pattern The pattern to check for.
   * @returns {number} The index of the pattern.
   */
  getRouteIndex(pattern) {
    for (let i = 0, j = this.state.routeStack.length; i < j; i += 1) {
      if (this.state.routeStack[i].pattern === pattern) {
        return i;
      }
    }
    return -1;
  }

  /**
   * Update the state based on the router changes.
   * @param {string} action The history action performed.
   * @param {string} currentPath The newly opened path.
   * @param {string} prevPath The previous path.
   * @param {string} routeStack The open routes.
   */
  handleRouteChange = () => {
    const routeStack = getRouteStack();
    this.setState({ routeStack });
  }

  /**
   * The render method.
   * @returns {JSX}
   */
  render() {
    /**
     * We need to apply some props to the children based on
     * the current router state.
     */
    // TODO: Move this code to another function.
    return React.Children.map(this.props.children, (route, index) => {

      // TODO: This is not the correct way to get the index.
      // Get the index of this route.
      const routeIndex = this.getRouteIndex(route.props.pattern);

      // Check if this route is open.
      const isOpen = routeIndex > -1;

      // Check if this route is the current visible route.
      const isVisible = routeIndex === this.state.routeStack.length - 1;

      // If this route is open then pass the pathname as a prop.
      const path = isOpen ? this.state.routeStack[routeIndex].pathname : null;
      const state = isOpen ? this.state.routeStack[routeIndex].state : null;

      return React.cloneElement(route, {
        index,
        isOpen,
        isVisible,
        path,
        state,
      });
    });
  }
}

export default Router;
