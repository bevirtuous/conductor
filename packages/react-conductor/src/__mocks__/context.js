export const RouteContext = 5;
export const RouterContext = 5;
