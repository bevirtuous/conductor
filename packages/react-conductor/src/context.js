import { createContext } from 'react';

export const RouteContext = createContext();
export const RouterContext = createContext();
export const SwitchContext = createContext();
