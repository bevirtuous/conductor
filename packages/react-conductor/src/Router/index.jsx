import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import conductor from '@virtuous/conductor';
import * as constants from '@virtuous/conductor/constants';
import * as events from '@virtuous/conductor-events';
import getCurrentAction from '@virtuous/conductor-helpers/getCurrentAction';
import getRouteStack from '@virtuous/conductor-helpers/getRouteStack';
import Route from '../Route';

const Context = React.createContext();
export const RouteContext = Context;

/**
 * The Router component.
 */
class Router extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
  };

  /**
   * The constructor.
   * @param {Object} props The component props.
   */
  constructor(props) {
    super(props);

    this.state = {
      stack: [],
    };

    // The route definitions.
    this.routes = {};

    // The stack of route definitions to use when rendering.
    this.routeStack = [];

    // Register the routes with Conductor.
    React.Children.forEach(props.children, child => this.registerChild(child.props));

    // Setup the callbacks for routing events.
    events.onDidPush(() => this.handleRouteChange(constants.ACTION_PUSH));
    events.onDidPop(() => this.handleRouteChange(constants.ACTION_POP));
    events.onDidReplace(() => this.handleRouteChange(constants.ACTION_REPLACE));
    events.onDidReset(() => this.handleRouteChange(constants.ACTION_RESET));
  }

  /**
   * @param {Object} nextProps The next props.
   * @param {Object} nextState The next state.
   * @returns {boolean}
   */
  shouldComponentUpdate(nextProps, nextState) {
    // Render if the stacks are different sizes.
    if (this.state.stack.length !== nextState.stack.length) {
      return true;
    }

    // Render if the history action was a replace.
    if (getCurrentAction() === constants.ACTION_REPLACE) {
      return true;
    }

    return false;
  }

  /**
   * Determines the last occurence of a path in the stack of open routes.
   * @param {Array} stack The open routes stack.
   * @param {string} pattern The pattern to look for.
   * @returns {number|null}
   */
  getlastOccurence = (stack, pattern) => {
    for (let i = stack.length - 1; i >= 0; --i) {
      if (stack[i].pattern === pattern) {
        return i;
      }
    }

    return null;
  }

  /**
   * Update the state based on the router changes.
   * TODO:
   */
  handleRouteChange = (action) => {
    const stack = getRouteStack();

    switch (action) {
      case constants.ACTION_PUSH: {
        //
        const { pattern } = stack[stack.length - 1];

        //
        if (this.routes[pattern].preload) {
          this.routeStack.push(this.routes[pattern].pattern);
        }

        break;
      }

      case constants.ACTION_POP: {
        //
        const { pattern } = this.state.stack[this.state.stack.length - 1];

        //
        if (this.routes[pattern].preload) {
          this.routeStack.pop();
        }

        break;
      }

      case constants.ACTION_REPLACE: {
        const { pattern } = this.state.stack[this.state.stack.length - 1];
        const newPattern = stack[stack.length - 1].pattern;

        //
        if (this.routes[pattern].preload) {
          this.routeStack.pop();
        }

        //
        if (this.routes[newPattern].preload) {
          this.routeStack.push(this.routes[newPattern].pattern);
        }

        break;
      }

      case constants.ACTION_RESET: {
        const { pattern, preload } = this.state.stack[0];

        // Set the routeStack back to it's initial state.
        this.routeStack.length = this.props.children.length;

        /**
         * Check if the initial entry is the stack is a preload route.
         * If so then immediately push another to the route stack.
         */
        if (preload) {
          this.routeStack.push(pattern);
        }

        break;
      }

      default:
        break;
    }

    this.setState({ stack });
  };

  /**
   * @param {Object} props - The child's props.
   */
  registerChild(props) {
    conductor.register(props.pattern);

    this.routes[props.pattern] = {
      component: props.component,
      pattern: props.pattern,
      preload: props.preload,
    };

    this.routeStack.push(props.pattern);
  }

  /**
   * TODO:
   * @returns {JSX}
   */
  renderOpenRoutes() {
    const ignoredSingletons = [];
    const usedRoutes = [];

    return (
      <Fragment>
        {this.routeStack.map((item, index) => {
          let pathname = null;
          let isVisible = false;
          let id = null;
          let params = {};
          let query = {};
          let state = {};

          const { component, pattern, preload } = this.routes[item];

          if (preload) {
            const match = this.state.stack.find((stackItem, stackItemIndex) => {
              const isMatch =
                stackItem.pattern === pattern && !usedRoutes.includes(stackItemIndex);

              if (isMatch) {
                usedRoutes.push(stackItemIndex);
              }

              if (isMatch && stackItemIndex === this.state.stack.length - 1) {
                isVisible = true;
              }

              return isMatch;
            });

            if (match) {
              id = match.id;
              pathname = match.pathname;
              params = match.params;
              query = match.query;
              state = match.state;
            }
          } else {
            // Skip this route if this pattern has already been handled.
            if (ignoredSingletons.includes(pattern)) {
              return null;
            }

            // We are handling this pattern only once, ignore it.
            ignoredSingletons.push(pattern);

            const lastOccurence = this.getlastOccurence(
              this.state.stack,
              pattern
            );

            if (this.state.stack[lastOccurence]) {
              id = this.state.stack[lastOccurence].id;
              pathname = this.state.stack[lastOccurence].pathname;
              params = this.state.stack[lastOccurence].params;
              query = this.state.stack[lastOccurence].query;
              state = this.state.stack[lastOccurence].state;
            }

            isVisible = lastOccurence === this.state.stack.length - 1;
          }

          const key = `route-${index}`;

          return (
            <Route
              id={id}
              component={component}
              isVisible={isVisible}
              key={key}
              params={params}
              path={pathname}
              pattern={pattern}
              query={query}
              state={state}
            />
          );
        })}
      </Fragment>
    );
  }

  /**
   * The render method.
   * @returns {JSX}
   */
  render() {
    return this.renderOpenRoutes();
  }
}

export default Router;
