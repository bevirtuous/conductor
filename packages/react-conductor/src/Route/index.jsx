import React from 'react';
import PropTypes from 'prop-types';
import { router } from '@virtuous/conductor';
import { RouteContext, RouterContext } from '../context';

/**
 * The Route component.
 */
class Route extends React.Component {
  static contextType = RouterContext;

  static propTypes = {
    component: PropTypes.func.isRequired,
    pattern: PropTypes.string.isRequired,
    transform: PropTypes.func,
  }

  static defaultProps = {
    transform: null,
  }

  /**
   * @param {Object} props The component props.
   */
  constructor(props) {
    super(props);
    router.register(props.pattern, props.transform);
  }

  /**
   * @returns {Object}
   */
  getContextValue = () => {
    const { ...context } = router.getCurrentRoute();
    context.open = true;
    context.visible = true;

    return context;
  }

  /**
   * @returns {JSX}
   */
  render() {
    const { component: Component, pattern } = this.props;
    const route = router.getCurrentRoute();

    if (route.pattern !== pattern) {
      return null;
    }

    const { transform, ...props } = route;
    const key = `${route.id}-${route.pathname}`;

    return (
      <RouteContext.Provider key={key} value={this.getContextValue()}>
        <Component route={props} />
      </RouteContext.Provider>
    );
  }
}

export default Route;
