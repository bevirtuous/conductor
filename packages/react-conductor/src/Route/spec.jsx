import React from 'react';
import { mount } from 'enzyme';
import { router } from '@virtuous/conductor';
import Router from '../Router/index';
import { RouteContext } from '../context';
import Route from './index';

const pattern = '/myroute/:id';
const spy = jest.spyOn(router, 'register');
const transform = route => route;

describe('<Route />', () => {
  it('should render component', () => {
    mount((
      <Router>
        <Route
          pattern={pattern}
          component={() => <div />}
          transform={transform}
        />
      </Router>
    ));

    expect(spy).toHaveBeenCalledWith(pattern, transform);
  });

  it('should correctly set the RouteContext value', () => {
    let contextValue = null;
    const current = router.getCurrentRoute();

    const MyComponent = () => (
      <RouteContext.Consumer>
        {(route) => {
          contextValue = route;
          return null;
        }}
      </RouteContext.Consumer>
    );

    mount((
      <Router>
        <Route
          pattern={pattern}
          component={MyComponent}
        />
      </Router>
    ));

    expect(current).toMatchObject(contextValue);
  });

  it('should render null when does not match current route', () => {
    const route = mount((
      <Router>
        <Route
          pattern="/wrong"
          component={() => <div />}
        />
      </Router>
    ));

    expect(route.html()).toBe('');
  });

  it('should react to router events and update', async () => {
    const route = mount((
      <Router>
        <Route
          pattern={pattern}
          component={() => <div />}
        />
        <Route
          pattern="/other"
          component={() => <div />}
        />
      </Router>
    ));

    // Should initially render the component (match found).
    expect(route.find(Route).at(0).html()).toBe('<div></div>');
    expect(route.find(Route).at(1).html()).toBeNull();

    await router.push({ pathname: '/other' });

    route.update();

    // Should now render null (first route does not match).
    expect(route.find(Route).at(0).html()).toBeNull();
    expect(route.find(Route).at(1).html()).toBe('<div></div>');
  });
});
