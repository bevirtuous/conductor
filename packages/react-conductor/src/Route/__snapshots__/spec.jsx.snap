// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Route /> should render component 1`] = `
<Router
  history={null}
>
  <Route
    component={[Function]}
    pattern="/myroute/:id"
    transform={[Function]}
  >
    <component
      route={
        Object {
          "created": 1556395116133,
          "hash": null,
          "id": "ruz2u",
          "location": "/myroute/123",
          "params": Object {
            "id": "123",
          },
          "pathname": "/myroute/123",
          "pattern": "/myroute/:id",
          "query": Object {},
          "state": Object {},
          "updated": null,
        }
      }
    >
      <div />
    </component>
  </Route>
</Router>
`;
